metapool/sample_sheet.py

```diff
@@ -101,45 +128,66 @@ class KLSampleSheet(sample_sheet.SampleSheet):
                 _BIOINFORMATICS_KEY, _CONTACT_KEY)
 
-    data_columns = (SS_SAMPLE_ID_KEY, _SS_SAMPLE_NAME_KEY, 'Sample_Plate',
-                    'Sample_Well', 'I7_Index_ID', 'index', 'I5_Index_ID',
-                    'index2', _SS_SAMPLE_PROJECT_KEY, 'Well_description')
-
-    column_alts = {'well_description': 'Well_description',
-                   'description': 'Well_description',
-                   'Description': 'Well_description',
-                   'sample_plate': 'Sample_Plate'}
-
-    CARRIED_PREP_COLUMNS = [EXPT_DESIGN_DESC_KEY, 'i5_index_id',
-                            'i7_index_id', 'index', 'index2',
-                            'library_construction_protocol',
-                            SAMPLE_NAME_KEY,
-                            'sample_plate', 'sample_project',
-                            'well_description', 'Sample_Well', 'Lane']
-
-    GENERATED_PREP_COLUMNS = ['center_name', 'center_project_name',
-                              'instrument_model', 'lane', 'platform',
-                              'run_center', 'run_date', 'run_prefix', 'runid',
-                              'sequencing_meth']
+    _ORDERED_BY_DATA_COLUMNS = False
+
+    # NB: Inside `make_sample_sheet`, the 'Well_description' column is
+    # (over)written by concatenating project plate, sample, and well.
+    # So it is required in the sense that it has to be present in the sample
+    # sheet, but it is not required in the sense that it has to be provided
+    # by a user when they create a sample sheet through this module.
+    _data_columns = (SS_SAMPLE_ID_KEY, _SS_SAMPLE_NAME_KEY, 'Sample_Plate',
+                     'Sample_Well', 'I7_Index_ID', 'index', 'I5_Index_ID',
+                     'index2', _SS_SAMPLE_PROJECT_KEY, 'Well_description')
+
+    _column_alts = MappingProxyType({'well_description': 'Well_description',
+                                     'description': 'Well_description',
+                                     'Description': 'Well_description',
+                                     'sample_plate': 'Sample_Plate'})
+
+    _CARRIED_PREP_COLUMNS = (EXPT_DESIGN_DESC_KEY, 'i5_index_id',
+                             'i7_index_id', 'index', 'index2',
+                             'library_construction_protocol',
+                             SAMPLE_NAME_KEY,
+                             'sample_plate', 'sample_project',
+                             'well_description', 'Sample_Well', 'Lane')
+
+    _GENERATED_PREP_COLUMNS = ('center_name', 'center_project_name',
+                               'instrument_model', 'lane', 'platform',
+                               'run_center', 'run_date', 'run_prefix', 'runid',
+                               'sequencing_meth')
+
+    @property
+    def CARRIED_PREP_COLUMNS(self):
+        return list(self._CARRIED_PREP_COLUMNS)
+
+    @property
+    def GENERATED_PREP_COLUMNS(self):
+        return list(self._GENERATED_PREP_COLUMNS)
```

Comment:
General question, what's the benefit of defining these variables as tuples and then transform them to lists?

Reply:
An excellent question!! Made even better because I actually have an answer :D 

`CARRIED_PREP_COLUMNS` and `GENERATED_PREP_COLUMNS` are among the rare sample sheet properties that actually get accessed from outside the sample sheet object.  I have not been able to find any places where the outside users try to *set* values in these properties (which is why I felt safe turning their underlying data structures into immutable tuples) BUT some of the outside users *are* using list comprehensions on these properties.  Python doesn't have tuple comprehensions, so the outside code fails if I just have the setter return the underlying tuple.  In order to keep this refactor functionally invisible to the outside users, I am therefore making a list out of the tuple before returning it so they can keep using it the way they always have.